File: twsapi_macunix/IBJts/samples/Cpp/TestCppClient/NewsSamplesProto.cpp
/* Copyright (C) 2026 Interactive Brokers LLC. All rights reserved. This code is subject to the terms
 * and conditions of the IB API Non-Commercial License or the IB API Commercial License, as applicable. */

#include "StdAfx.h"

#include "NewsSamplesProto.h"

#if !defined(USE_WIN_DLL)

protobuf::HistoricalNewsRequest NewsSamplesProto::HistoricalNewsRequestWithEndTime(int reqId){
    //! [historical_news_request_with_end_time]
    protobuf::HistoricalNewsRequest historicalNewsRequestProto;
    historicalNewsRequestProto.set_reqid(reqId);
    historicalNewsRequestProto.set_conid(8314);
    historicalNewsRequestProto.set_providercodes("BRFUPDN+BRFG");
    time_t now = time(nullptr);
    now -= 10 * 24 * 60 * 60; // subtract 10 days in seconds
    struct tm timeinfo;
#if defined(IB_WIN32)
    localtime_s(&timeinfo, &now);
#else
    localtime_r(&now, &timeinfo);
#endif
    char buffer[80];
    strftime(buffer, sizeof(buffer), "%Y-%m-%d %H:%M:%S", &timeinfo);
    historicalNewsRequestProto.set_enddatetime(std::string(buffer) + ".0");
    historicalNewsRequestProto.set_totalresults(10);
    //! [historical_news_request_with_end_time]
    return historicalNewsRequestProto;
}

protobuf::HistoricalNewsRequest NewsSamplesProto::HistoricalNewsRequestWithStartTime(int reqId){
    //! [historical_news_request_with_start_time]
    protobuf::HistoricalNewsRequest historicalNewsRequestProto;
    historicalNewsRequestProto.set_reqid(reqId);
    historicalNewsRequestProto.set_conid(8314);
    historicalNewsRequestProto.set_providercodes("BRFUPDN+BRFG");
    time_t now = time(nullptr);
    now -= 10 * 24 * 60 * 60; // subtract 10 days in seconds
    struct tm timeinfo;
#if defined(IB_WIN32)
    localtime_s(&timeinfo, &now);
#else
    localtime_r(&now, &timeinfo);
#endif
    char buffer[80];
    strftime(buffer, sizeof(buffer), "%Y-%m-%d %H:%M:%S", &timeinfo);
    historicalNewsRequestProto.set_startdatetime(std::string(buffer) + ".0");
    historicalNewsRequestProto.set_totalresults(10);
    //! [historical_news_request_with_start_time]
    return historicalNewsRequestProto;
}

#endif

File: twsapi_macunix/IBJts/samples/Python/Testbed/NewsSamplesProto.py
"""
Copyright (C) 2026 Interactive Brokers LLC. All rights reserved. This code is subject to the terms
 and conditions of the IB API Non-Commercial License or the IB API Commercial License, as applicable.
"""

from ibapi.protobuf.HistoricalNewsRequest_pb2 import HistoricalNewsRequest as HistoricalNewsRequestProto
from datetime import datetime, timedelta

class NewsSamplesProto:

    @staticmethod
    def HistoricalNewsRequestWithEndTime(reqId: int) -> HistoricalNewsRequestProto:
        #! [historical_news_request_with_end_time]
        historicalNewsRequestProto = HistoricalNewsRequestProto()
        historicalNewsRequestProto.reqId = reqId
        historicalNewsRequestProto.conId = 8314
        historicalNewsRequestProto.providerCodes = "BRFUPDN+BRFG"
        date_10_days_ago = datetime.now() - timedelta(days=10)
        historicalNewsRequestProto.endDateTime = date_10_days_ago.strftime("%Y-%m-%d %H:%M:%S.0")
        historicalNewsRequestProto.totalResults = 10
        #! [historical_news_request_with_end_time]
        return historicalNewsRequestProto

    @staticmethod
    def HistoricalNewsRequestWithStartTime(reqId: int) -> HistoricalNewsRequestProto:
        #! [historical_news_request_with_start_time]
        historicalNewsRequestProto = HistoricalNewsRequestProto()
        historicalNewsRequestProto.reqId = reqId
        historicalNewsRequestProto.conId = 8314
        historicalNewsRequestProto.providerCodes = "BRFUPDN+BRFG"
        date_10_days_ago = datetime.now() - timedelta(days=10)
        historicalNewsRequestProto.startDateTime = date_10_days_ago.strftime("%Y-%m-%d %H:%M:%S.0")
        historicalNewsRequestProto.totalResults = 10
        #! [historical_news_request_with_start_time]
        return historicalNewsRequestProto

def Test():
    from ibapi.utils import ExerciseStaticMethods
    ExerciseStaticMethods(NewsSamplesProto)

if "__main__" == __name__:
    Test()

